### PR TITLE
Em/transient

### PR DESCRIPTION
**Pull request by APaganini**
New `PDEconstrainedObjective` class to support transient objective functions.
`ReducedObjective` is no longer required and is flagged as `Deprecated`.

**Note**: Since we rely on `pyadjoint` to differentiate `ReducedObjective`s, this class
does not use `J.value_form()`, but only `J.value()`. Therefore, `ShapeObjectives` in
`ReducedObjectives` are morally treated differently from standalone `ShapeObjectives`.

### CODE REVIEW

**Comment by APaganini** [comment]
I'm confused by the test failure on circleci. It currently runs on my machine.

**Update:** It's a problem with the tape. Sorting it out!

File: fireshape/objective.py [hunk 1]
```diff
@@ -212,7 +284,15 @@ def update(self, x, flag, iteration):
...
-                fda.continue_annotation()
+                # ensure we are annotating
+                from pyadjoint.tape import annotate_tape
+                safety_counter = 0
+                while not annotate_tape():
+                    safety_counter += 1
+                    fda.continue_annotation()
+                    if safety_counter > 1e2:
+                        import sys
+                        sys.exit('Cannot annotating even after 100 attempts.')
```
Comment (florianwechsung) [inline fireshape/objective.py]: annotating -> annotate

File: tests/test_TimeTracking.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+            t += self.dt
+            self.J += fd.assemble(self.dt*(self.u - self.u_t(t))**2*self.dx)
+
+    def objective_value(self):
+        """Return the value of the objective function."""
+        return self.J
+
+
+def test_TimeTracking():
+    """ Test template for fsz.LevelsetFunctional."""
```
Comment (florianwechsung) [inline tests/test_TimeTracking.py]: This sentence seems misplaced?

**Comment by florianwechsung** [comment]
Looks good to me, just two minor comments, then happy to merge.

File: fireshape/objective.py (hunk 1 above)
Comment (APaganini) [reply fireshape/objective.py]: done

File: tests/test_TimeTracking.py (hunk 2 above)
Comment (APaganini) [reply tests/test_TimeTracking.py]: fixed

**Review by florianwechsung — APPROVED** [review]

**Pull request merged**